REGRESSION(274311@main) YouTube.com in safari Airplay connects to ATV receiver, no Video plays but only Audio

jyavenard opened the pull request:
#### d9ce6b4272d071babd1ff0ee646056cfb06f8b37
<pre>
REGRESSION(274311@main) YouTube.com in safari Airplay connects to ATV receiver, no Video plays but only Audio
<a href="https://bugs.webkit.org/show_bug.cgi?id=270437">https://bugs.webkit.org/show_bug.cgi?id=270437</a>
<a href="https://rdar.apple.com/123467550">rdar://123467550</a>

Reviewed by Youenn Fablet.

When activating an AirPlay target on a currently playing media element,
the UI process was sending a MediaPlaybackTargetContext to the web process, which
would then pass it to the GPUP&apos;s AVPlayer.
Following 274311@main, the AVOutputContext found in the MediaPlaybackTargetContext
would be serialized/deserialized at each IPC&apos;s endpoints. This relied on the
generic AVOutputContext IPC serialization introduced in 271664@main.
However, re-constructing an AVOutputContext in the WP is not allowed by
the sandbox. An AVOutputContext&apos;s serialization is made of a global identification
string which can be looked up in a global table. This look-uo would always
fail in the WP resulting in an null AVOutputContext being generated.
Subsequently, the GPUP would receive this null AVOutputContext causing the
AirPlay command to never reach the GPUP.

On iOS devices similarly, the GPUP sends the notification that a new AVOutputContext had
been selected, sends it to the WP which then re-send it for the GPUP for processing.
As above, the transit through the WP would null the AVOutputContext.

We revert to the behaviour pre-274311@main which would use a transient SerializedAVOutputContext
to pass a serialized AVOutputContext, with a cache of hasActiveRoute which would
only be reconstructed in the GPUP.
In order to keep the security aspect intended by 274311@main we introduce two
new classes: MediaPlaybackTargetSerialized and MediaPlaybackTargetContextSerialized
which will wrap either a MediaPlaybackTargetContextCocoa or a MediaPlaybackTargetContextMock.
The underlying AVOutputContext is now never accessed in the WP. Instead we only pass
the underlying AVOutputContext&apos;s global ID.
We split MediaPlaybackTargetContext into three sub-classes:
- Cocoa/AVOutputContext
- Mock
- Serialized

The MediaPlaybackTargetSerialized now uses the new serialization framework.

Manually tested to work with an ATV and a macOS sonoma as AirPlay target.
Some work covered by MediaPlaybackTargetMock tests; but infrastructure doesn&apos;t
allow for comprehensive regression tests.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/MediaPlaybackTarget.h:
* Source/WebCore/platform/graphics/avfoundation/MediaPlaybackTargetCocoa.h:
(isType):
* Source/WebCore/platform/graphics/avfoundation/MediaPlaybackTargetCocoa.mm:
(WebCore::MediaPlaybackTargetContextCocoa::MediaPlaybackTargetContextCocoa):
(WebCore::MediaPlaybackTargetContextCocoa::deviceName const):
(WebCore::MediaPlaybackTargetContextCocoa::hasActiveRoute const):
(WebCore::MediaPlaybackTargetContextCocoa::supportsRemoteVideoPlayback const):
(WebCore::MediaPlaybackTargetCocoa::create):
(WebCore::MediaPlaybackTargetCocoa::MediaPlaybackTargetCocoa):
(WebCore::MediaPlaybackTargetCocoa::~MediaPlaybackTargetCocoa): Deleted.
(WebCore::toMediaPlaybackTargetCocoa): Deleted.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlaybackTargetPickerMac.mm:
(WebCore::MediaPlaybackTargetPickerMac::playbackTarget):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::setWirelessPlaybackTarget):
* Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
(WebCore::MediaPlaybackTargetContext::MediaPlaybackTargetContext):
(WebCore::MediaPlaybackTargetContext::mockState const): Deleted.
(WebCore::MediaPlaybackTargetContext::outputContext const): Deleted.
(WebCore::MediaPlaybackTargetContext::encodingRequiresPlatformData const): Deleted.
(): Deleted.
* Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.mm: Removed.
* Source/WebCore/platform/mock/MediaPlaybackTargetMock.cpp:
(WebCore::MediaPlaybackTargetMock::create):
(WebCore::MediaPlaybackTargetMock::MediaPlaybackTargetMock):
(WebCore::toMediaPlaybackTargetMock): Deleted.
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h: Add missing forward declaration.
* Source/WebCore/platform/mock/MediaPlaybackTargetMock.h:
(isType):
(WebCore::MediaPlaybackTargetMock::state const): Deleted.
* Source/WebCore/platform/mock/MediaPlaybackTargetPickerMock.cpp:
(WebCore::MediaPlaybackTargetPickerMock::playbackTarget):
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::setWirelessPlaybackTarget):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.messages.in:
* Source/WebKit/GPUProcess/media/ios/RemoteMediaSessionHelperProxy.cpp:
(WebKit::RemoteMediaSessionHelperProxy::activeVideoRouteDidChange):
* Source/WebKit/Platform/SourcesCocoa.txt:
* Source/WebKit/Platform/cocoa/MediaPlaybackTargetContextSerialized.h: Added.
(isType):
* Source/WebKit/Platform/cocoa/MediaPlaybackTargetContextSerialized.mm: Added.
(WebKit::MediaPlaybackTargetContextSerialized::MediaPlaybackTargetContextSerialized):
(WebKit::MediaPlaybackTargetContextSerialized::platformContext const):
* Source/WebKit/Platform/cocoa/MediaPlaybackTargetContextSerialized.serialization.in: Renamed from Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.serialization.in.
* Source/WebKit/Scripts/webkit/messages.py:
(conditions_for_header):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setMockMediaPlaybackTargetPickerState):
(WebKit::WebPageProxy::Internals::setPlaybackTarget):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::setWirelessPlaybackTarget):
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.cpp:
(WebKit::RemoteMediaSessionHelper::activeVideoRouteDidChange):
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.h:
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::playbackTargetSelected const):

Canonical link: <a href="https://commits.webkit.org/275730@main">https://commits.webkit.org/275730@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/74ef224604b862089880f6dafcc8e6cb8f71b898

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42658 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21680 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45060 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45269 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38782 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44965 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/25342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19045 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35308 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43232 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/25342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36715 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16262 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/42529 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/25342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37774 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/715 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/25342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38104 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46767 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/17477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14402 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42020 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/19096 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40637 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9531 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19275 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18741 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/avfoundation/MediaPlaybackTargetCocoa.h:
```diff
@@ -28,34 +28,38 @@
+    AVOutputContext* outputContext() { return m_context.outputContext().get(); }
+
 protected:
-    explicit MediaPlaybackTargetCocoa(AVOutputContext *);
```
Why protected, can we make all of them private?
Should MediaPlaybackTargetCocoa be final?

jyavenard replied:
we can; visibility was all existing code. I don't tend to modif what was already in place like that.

youennf commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
```diff
@@ -42,62 +38,31 @@ enum class MediaPlaybackTargetContextMockState : uint8_t {
-        : m_type(Type::Mock)
-        , m_mockDeviceName(mockDeviceName)
-        , m_mockState(state)
+    MediaPlaybackTargetContext(Type type)
```
explicit

youennf commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContextCocoa.h:
```diff
@@ -0,0 +1,65 @@
+
+namespace WebCore {
+
+class MediaPlaybackTargetContextCocoa : public MediaPlaybackTargetContext {
```
final?

youennf commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContextCocoa.h:
```diff
@@ -0,0 +1,65 @@
+
+class MediaPlaybackTargetContextCocoa : public MediaPlaybackTargetContext {
+public:
+    WEBCORE_EXPORT MediaPlaybackTargetContextCocoa(RetainPtr<AVOutputContext>&&);
```
explicit

jyavenard replied:
see below, making an explicit operator render the construction of a std::variant impossible. 

https://en.cppreference.com/w/cpp/utility/variant/variant :
4) Converting constructor. Constructs a variant holding the alternative type T_j that would be selected by overload resolution for the expression F([std::forward](http://en.cppreference.com/w/cpp/utility/forward)<T>(t)) if there was an overload of imaginary function F(T_i) for every T_i from Types... in scope at the same time, except that:
An overload F(T_i) is only considered if the declaration T_i x[] = { [std::forward](http://en.cppreference.com/w/cpp/utility/forward)<T>(t) }; is valid for some invented variable x.

youennf commented on Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContextCocoa.h:
```diff
@@ -0,0 +1,65 @@
+public:
+    WEBCORE_EXPORT MediaPlaybackTargetContextCocoa(RetainPtr<AVOutputContext>&&);
+    WEBCORE_EXPORT ~MediaPlaybackTargetContextCocoa();
+    WEBCORE_EXPORT MediaPlaybackTargetContextCocoa(const MediaPlaybackTargetContextCocoa&);
```
We usually define operator= in case of copy constructor.

jyavenard replied:
Those operators shouldn’t be used. I had to define them and make them external as otherwise the IPC serialiser caused linkage error with missing vtable

jyavenard replied:
somehow it now compiles without them after I exported the base class default constructor/destructor.

jyavenard replied:
spoke too fast, it is still happening on Mac build :(

jyavenard replied:
```4) Converting constructor. Constructs a variant holding the alternative type T_j that would be selected by overload resolution for the expression F([std::forward](http://en.cppreference.com/w/cpp/utility/forward)<T>(t)) if there was an overload of imaginary function F(T_i) for every T_i from Types... in scope at the same time, except that:
An overload F(T_i) is only considered if the declaration T_i x[] = { [std::forward](http://en.cppreference.com/w/cpp/utility/forward)<T>(t) }; is valid for some invented variable x.
```

https://godbolt.org/z/hW6v5oYxf

to be able to use a `MediaPlaybackTargetContextCocoa` in a std::variant, you can't have an explicit constructor.

youennf commented on Source/WebCore/platform/mock/MediaPlaybackTargetContextMock.h:
```diff
@@ -0,0 +1,68 @@
+
+namespace WebCore {
+
+class MediaPlaybackTargetContextMock : public MediaPlaybackTargetContext {
```
final

youennf commented on Source/WebCore/platform/mock/MediaPlaybackTargetContextMock.h:
```diff
@@ -0,0 +1,68 @@
+
+    String deviceName() const final { return m_mockDeviceName; }
+    bool hasActiveRoute() const final { return !m_mockDeviceName.isEmpty(); }
+    bool supportsRemoteVideoPlayback() const final { return !m_mockDeviceName.isEmpty(); }
```
private?

jyavenard replied:
the class is only accessed via `MediaPlaybackTargetMock` and not through the base class.

youennf commented on Source/WebCore/platform/mock/MediaPlaybackTargetMock.h:
```diff
@@ -25,33 +25,38 @@
-    MediaPlaybackTargetContext::MockState state() const { return m_context.mockState(); }
+    MediaPlaybackTargetContextMock::State state() const { return m_context.state(); }
 
 protected:
```
private and make class final?

youennf commented on Source/WebKit/Platform/cocoa/MediaPlaybackTargetSerialized.h:
```diff
@@ -0,0 +1,63 @@
+
+namespace WebKit {
+
+class MediaPlaybackTargetSerialized : public WebCore::MediaPlaybackTarget {
```
final

youennf commented on Source/WebKit/Platform/cocoa/MediaPlaybackTargetSerialized.h:
```diff
@@ -0,0 +1,63 @@
+    const WebCore::MediaPlaybackTargetContext& targetContext() const final { return m_context; }
+
+private:
+    MediaPlaybackTargetSerialized(MediaPlaybackTargetContextSerialized&& context)
```
explicit

jyavenard committed:
REGRESSION(274311@main) YouTube.com in safari Airplay connects to ATV receiver, no Video plays but only Audio
https://bugs.webkit.org/show_bug.cgi?id=270437
rdar://123467550

Reviewed by Youenn Fablet.

When activating an AirPlay target on a currently playing media element,
the UI process was sending a MediaPlaybackTargetContext to the web process, which
would then pass it to the GPUP's AVPlayer.
Following 274311@main, the AVOutputContext found in the MediaPlaybackTargetContext
would be serialized/deserialized at each IPC's endpoints. This relied on the
generic AVOutputContext IPC serialization introduced in 271664@main.
However, re-constructing an AVOutputContext in the WP is not allowed by
the sandbox. An AVOutputContext's serialization is made of a global identification
string which can be looked up in a global table. This look-uo would always
fail in the WP resulting in an null AVOutputContext being generated.
Subsequently, the GPUP would receive this null AVOutputContext causing the
AirPlay command to never reach the GPUP.

On iOS devices similarly, the GPUP sends the notification that a new AVOutputContext had
been selected, sends it to the WP which then re-send it for the GPUP for processing.
As above, the transit through the WP would null the AVOutputContext.

We revert to the behaviour pre-274311@main which would use a transient SerializedAVOutputContext
to pass a serialized AVOutputContext, with a cache of hasActiveRoute which would
only be reconstructed in the GPUP.
In order to keep the security aspect intended by 274311@main we introduce two
new classes: MediaPlaybackTargetSerialized and MediaPlaybackTargetContextSerialized
which will wrap either a MediaPlaybackTargetContextCocoa or a MediaPlaybackTargetContextMock.
The underlying AVOutputContext is now never accessed in the WP. Instead we only pass
the underlying AVOutputContext's global ID.
We split MediaPlaybackTargetContext into three sub-classes:
- Cocoa/AVOutputContext
- Mock
- Serialized

The MediaPlaybackTargetSerialized now uses the new serialization framework.

Manually tested to work with an ATV and a macOS sonoma as AirPlay target.
Some work covered by MediaPlaybackTargetMock tests; but infrastructure doesn't
allow for comprehensive regression tests.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/MediaPlaybackTarget.h:
* Source/WebCore/platform/graphics/avfoundation/MediaPlaybackTargetCocoa.h:
(isType):
* Source/WebCore/platform/graphics/avfoundation/MediaPlaybackTargetCocoa.mm:
(WebCore::MediaPlaybackTargetContextCocoa::MediaPlaybackTargetContextCocoa):
(WebCore::MediaPlaybackTargetContextCocoa::deviceName const):
(WebCore::MediaPlaybackTargetContextCocoa::hasActiveRoute const):
(WebCore::MediaPlaybackTargetContextCocoa::supportsRemoteVideoPlayback const):
(WebCore::MediaPlaybackTargetCocoa::create):
(WebCore::MediaPlaybackTargetCocoa::MediaPlaybackTargetCocoa):
(WebCore::MediaPlaybackTargetCocoa::~MediaPlaybackTargetCocoa): Deleted.
(WebCore::toMediaPlaybackTargetCocoa): Deleted.
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlaybackTargetPickerMac.mm:
(WebCore::MediaPlaybackTargetPickerMac::playbackTarget):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::setWirelessPlaybackTarget):
* Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.h:
(WebCore::MediaPlaybackTargetContext::MediaPlaybackTargetContext):
(WebCore::MediaPlaybackTargetContext::mockState const): Deleted.
(WebCore::MediaPlaybackTargetContext::outputContext const): Deleted.
(WebCore::MediaPlaybackTargetContext::encodingRequiresPlatformData const): Deleted.
(): Deleted.
* Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.mm: Removed.
* Source/WebCore/platform/mock/MediaPlaybackTargetMock.cpp:
(WebCore::MediaPlaybackTargetMock::create):
(WebCore::MediaPlaybackTargetMock::MediaPlaybackTargetMock):
(WebCore::toMediaPlaybackTargetMock): Deleted.
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h: Add missing forward declaration.
* Source/WebCore/platform/mock/MediaPlaybackTargetMock.h:
(isType):
(WebCore::MediaPlaybackTargetMock::state const): Deleted.
* Source/WebCore/platform/mock/MediaPlaybackTargetPickerMock.cpp:
(WebCore::MediaPlaybackTargetPickerMock::playbackTarget):
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::setWirelessPlaybackTarget):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.messages.in:
* Source/WebKit/GPUProcess/media/ios/RemoteMediaSessionHelperProxy.cpp:
(WebKit::RemoteMediaSessionHelperProxy::activeVideoRouteDidChange):
* Source/WebKit/Platform/SourcesCocoa.txt:
* Source/WebKit/Platform/cocoa/MediaPlaybackTargetContextSerialized.h: Added.
(isType):
* Source/WebKit/Platform/cocoa/MediaPlaybackTargetContextSerialized.mm: Added.
(WebKit::MediaPlaybackTargetContextSerialized::MediaPlaybackTargetContextSerialized):
(WebKit::MediaPlaybackTargetContextSerialized::platformContext const):
* Source/WebKit/Platform/cocoa/MediaPlaybackTargetContextSerialized.serialization.in: Renamed from Source/WebCore/platform/graphics/cocoa/MediaPlaybackTargetContext.serialization.in.
* Source/WebKit/Scripts/webkit/messages.py:
(conditions_for_header):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setMockMediaPlaybackTargetPickerState):
(WebKit::WebPageProxy::Internals::setPlaybackTarget):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/media/MediaPlayerPrivateRemote.cpp:
(WebKit::MediaPlayerPrivateRemote::setWirelessPlaybackTarget):
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.cpp:
(WebKit::RemoteMediaSessionHelper::activeVideoRouteDidChange):
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.h:
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::playbackTargetSelected const):

Canonical link: https://commits.webkit.org/275730@main
